compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/VectorizeIREEVectorExtOps.cpp

```diff
@@ -549,14 +569,56 @@ vectorizeLinalgExtGatherToTransferGather(RewriterBase &rewriter,
       rewriter.getMultiDimIdentityMap(gatherTy.getRank()), padding,
       /*mask=*/Value(), inBounds);
 
+  VectorType gatherMaskType = gatherVectorTy.clone(rewriter.getI1Type());
+  Value gatherMask =
+      rewriter.create<vector::CreateMaskOp>(loc, gatherMaskType, gatherDims);
+  Value maskedGather = maskOperation(rewriter, transferGatherOp, gatherMask);
+
   // Write back into tensor.
-  auto emptyOp = rewriter.create<tensor::EmptyOp>(loc, gatherTy.getShape(),
-                                                  gatherTy.getElementType());
   SmallVector<Value> writeIndices(gatherTy.getRank(), zero);
   auto writeOp = rewriter.create<vector::TransferWriteOp>(
-      loc, transferGatherOp.getResult(), emptyOp, writeIndices);
-  rewriter.replaceOp(gatherOp, writeOp);
+      loc, maskedGather, gatherOp.getOutput(), writeIndices);
+  Value maskedWrite = maskOperation(rewriter, writeOp, gatherMask);
+
+  rewriter.replaceOp(gatherOp, maskedWrite);
   return success();
 }
 
+/// Lowers vector.mask %mask { iree_vector_ext.transfer_gather }
+///  into
+/// iree_vector_ext.transfer_gather %mask
+///
+/// Ideally, the mask should have just been put on transfer_gather directly,
+/// but this is done this way to match upstream vector.transfer_read masking.
+struct MaskedTransferGatherOpPattern : public OpRewritePattern<vector::MaskOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::MaskOp maskOp,
+                                PatternRewriter &rewriter) const override {
+    auto gatherOp = dyn_cast<TransferGatherOp>(maskOp.getMaskableOp());
+    if (!gatherOp) {
+      return failure();
+    }
+    // TODO: The 'vector.mask' passthru is a vector and 'vector.transfer_read'
+    // expects a scalar. We could only lower one to the other for cases where
+    // the passthru is a broadcast of a scalar.
+    if (maskOp.hasPassthru()) {
+      return rewriter.notifyMatchFailure(
+          maskOp, "can't lower passthru to transfer_gather");
+    }
+    // Replace the `vector.mask` operation.
```

Comment:
I'd remove the comment because the code is self-documented. I.e., it explicitly says `replaceOpWithNewOp<TransferGatherOp>(maskOp ...`